content/topics/Analyze/machine-learning/supervised/K-fold_crossvalidation.md

```diff
@@ -0,0 +1,103 @@
+
+5- Hyperparameter tuning: By finding the best combination of hyperparameters, the model's complexity can be managed. Examples of hyperparameters include the learning rate in gradient descent and the depth of a decision tree. Overall, hyperparameter tuning helps a balance between model complexity and generalization performance, ultimately reducing the risk of overfitting.
+
+## K-Fold Cross-Validation
+One method to check how well a model works on new data is using train-test-split. Here, we set aside some data to test the model after training it. But a better method is K-fold cross-validation. In this technique, we split the dataset into k subsets or folds. The model is trained k times, each time using a different fold for testing and the rest for training. This gives us k evaluation scores. Finally, we average these scores to get the overall performance. This approach provides a more reliable evaluation compared to just using train-test-split.
+Using K-fold cross-validation doesn't directly prevent overfitting, but it aids in detecting and addressing it. Let's explore it further by means of an example.
+
+We'll use the diabetes dataset from Sklearn and train a DecisionTreeRegressor to predict diabetes progression.
+{{% tip %}}
+Tree-based algorithms are particularly vulnerable to overfitting. 
+{{% /tip %}}
+
+Running the next code block allows you to split the dataset into training and testing sets. You then train the model using the training set and assess its performance by calculating the Mean Squared Error (MSE) for both the training and testing datasets.
+{{% codeblock %}}
+```Python
+from sklearn.datasets import load_diabetes
+from sklearn.model_selection import train_test_split, cross_val_score
+from sklearn.tree import DecisionTreeRegressor
+from sklearn.metrics import mean_squared_error
+
+# Load the diabetes dataset
+diabetes = load_diabetes()
+X = diabetes.data
+y = diabetes.target
+
+# a simple train-test split 
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
+
+# training the model
+model = DecisionTreeRegressor()
+model.fit(X_train, y_train)
+
+# Make prediction
+y_train_pred = model.predict(X_train)
+y_test_pred = model.predict(X_test)
+
+# Calculate Mean Squared Error (MSE)
+mse_train = mean_squared_error(y_train, y_train_pred)
+mse_test = mean_squared_error(y_test, y_test_pred)
+
+print("MSE for Training Data:", mse_train)
+print("MSE for Test Data:", mse_test)
+```
+{{% /codeblock %}}
+
+The result shows that the training MSE is zero, whereas the test MSE is 4933, indicating a severe case of overfitting. We can see this because the MSE in test set is **much** larger than in the training set. When we  implement K-fold cross-validation in next codeblock, it becomes even evident that the model is overfits the training data.
```

Comment:
```suggestion
The result shows that the training MSE is zero, whereas the test MSE is `4933`, indicating a severe case of overfitting. We can see this because the MSE in the test set is **much** larger than in the training set. When we  implement K-fold cross-validation in the next code block, it becomes even evident that the model overfits the training data.
```